ls1intum/Artemis
# Add support for displaying human-readable breadcrumb labels instead of IDs

### Checklist
- [x] I tested *all* changes and *all* related features on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
Currently breadcrumb labels are just the raw urls, not really user readable. (E.g. the course ID instead of the name.)

### Description
To properly solve this issue, routes need to be re-factored to use children instead of absolute paths. This is a lot of work, so the PR here just starts on a smaller scale (completely porting System Administration, but only parts of the Course Administration). Yet this PR is still pretty big, so I suggest **reviewing the code commit by commit**.

How routes would ideally look using children (and do on this PR):
```
course management
├───create
├───course id
│   ├───update
│   ├───scores
│   ├───exercises
│   │   ├───create
│   │   ├───id
│   │   └───...
│   ├───lectures
│   │   ├───create
│   │   ├───id
│   │   └───...
```
How they more or less look on `develop`:
```
course management
├───create
├───course id
├───course id───update
├───course id───scores
├───course id───exercises
├───course id───exercises───create
├───course id───exercises───id
├───course id───lectures
├───course id───lectures───create
├───course id───lectures───id
├───...
```

The reason fixing this is taking so long is that several components do not use resolvers, but implemented their own assumptions about the routes they are in, so those need to be updated as well. Then there is also the problem with routes being inconsistent and - even more annoyingly - spread out over a lot of files which makes it really hard to get an overview. Lazily loaded routes also can't access routes outside of their root component, which makes sense but means I had to put in a workaround to manually define breadcrumbs for such cases. (This will mostly be used in follow-up PRs since I didn't touch most of the lazily loaded routes.)

### Steps for Testing

1. Log in to Artemis
2. Test all buttons that move you to another page from the Course or System Administration
3. Check that the breadcrumb labels in the System Administration paths are human readable
4. Check that all other breadcrumbs are either unchanged (raw url) or readable
5. Check that the crumbs point to the correct paths

### Screenshots

Breadcrumbs now use human readable labels in several places:
![grafik](https://user-images.githubusercontent.com/72132281/100517039-45207380-3188-11eb-8596-1313fec60fa8.png)
![grafik](https://user-images.githubusercontent.com/72132281/100517059-697c5000-3188-11eb-8d4e-b8f2d2d98fb4.png)

A lot of components still use the old behaviour until their routes have been refactored:
![grafik](https://user-images.githubusercontent.com/72132281/100517049-58cbda00-3188-11eb-882a-157a4db73677.png)
